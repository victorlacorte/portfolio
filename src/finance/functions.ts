import * as R from 'ramda';
import { add, div, mul, sub, trunc } from '../utils/number';

type Base = (price: number, quantity: number, tax: number) => number;
type Profit = (
  ...params: [...base: Parameters<Base>, prevPrice: number]
) => number;

const truncate = (x: number): number => trunc(x, 4);

function _averagePrice(price: number, quantity: number, tax: number): number;
function _averagePrice(
  price: number,
  quantity: number,
  tax: number,
  prevPrice: number,
  prevQuantity: number,
): number;
function _averagePrice(
  price: number,
  quantity: number,
  tax: number,
  prevPrice?: number,
  prevQuantity?: number,
): number {
  const total = _buyTotal(price, quantity, tax);

  if (prevPrice === undefined && prevQuantity === undefined) {
    return div(total, quantity);
  }

  return div(
    add(mul(prevPrice, prevQuantity), total),
    add(prevQuantity, quantity),
  );
}

const _buyTotal: Base = (price, quantity, tax): number =>
  add(mul(price, quantity), tax);

const _sellTotal: Base = (price, quantity, tax): number =>
  sub(mul(price, quantity), tax);

const _profit: Profit = (price, quantity, tax, prevPrice) =>
  _sellTotal(sub(price, prevPrice), quantity, tax);

const _profitPercent: Profit = (price, quantity, tax, prevPrice) =>
  sub(div(_sellTotal(price, quantity, tax), mul(prevPrice, quantity)), 1);

export const averagePrice: typeof _averagePrice = R.pipe(
  _averagePrice,
  truncate,
);

export const buyTotal: typeof _buyTotal = R.pipe(_buyTotal, truncate);

export const sellTotal: typeof _sellTotal = R.pipe(_sellTotal, truncate);

export const profit: typeof _profit = R.pipe(_profit, truncate);

export const profitPercent: typeof _profitPercent = R.pipe(
  _profitPercent,
  truncate,
);

// TODO these functions are simple date utilities
// type CompareFn = (a: PositionEntry, b: PositionEntry) => number;
// const compareAscending: CompareFn = (a, b) =>
//   a.date.equals(b.date)
//     ? -1
//     : Number(a.date.toJSDate()) - Number(b.date.toJSDate());
// const compareDescending: CompareFn = (a, b) =>
//   b.date.equals(a.date)
//     ? 1
//     : Number(b.date.toJSDate()) - Number(a.date.toJSDate());
