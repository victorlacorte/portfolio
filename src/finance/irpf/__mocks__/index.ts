import SimpleDate from 'src/utils/date';
import type {
  BaseEntry,
  BuyEntry,
  PositionEntry,
  SellEntry,
  SimpleDate as _SimpleDate,
} from 'src/types';

type SortByDate = (entries: PositionEntry[]) => PositionEntry[];

type EntryProperties<T> = Partial<Omit<BaseEntry, 'date'> & T & YMD>;
type MakeEntry<T> = (properties?: EntryProperties<T>) => T & BaseEntry;

type MakeBuyEntry = MakeEntry<BuyEntry>;
type MakeSellEntry = MakeEntry<SellEntry>;

type YMD = {
  year: number;
  month: number;
  day: number;
};

// The year itself does not matter but will dictate all calculations
export const year = 2020;

const defaultBuyEntry: Required<EntryProperties<BuyEntry>> = {
  year: 2020,
  month: 1,
  day: 1,
  price: 1,
  quantity: 1,
  buyQuantity: 100,
  buyTotal: 100,
};

const defaultSellEntry: Required<EntryProperties<SellEntry>> = {
  year: 2020,
  month: 1,
  day: 1,
  price: 1,
  quantity: 1,
  sellQuantity: 100,
  sellIrrf: 1,
  sellProfit: 1,
  sellProfitPercent: 0.1,
  sellTotal: 100,
};

export const makeBuyEntry: MakeBuyEntry = ({
  year,
  month,
  day,
  ...properties
} = {}) => ({
  ...defaultBuyEntry,
  ...properties,
  date: SimpleDate.make(
    year ?? defaultBuyEntry.year,
    month ?? defaultBuyEntry.month,
    day ?? defaultBuyEntry.day,
  ),
});

export const makeSellEntry: MakeSellEntry = ({
  year,
  month,
  day,
  ...properties
} = {}) => ({
  ...defaultSellEntry,
  ...properties,
  date: SimpleDate.make(
    year ?? defaultBuyEntry.year,
    month ?? defaultBuyEntry.month,
    day ?? defaultBuyEntry.day,
  ),
});

export const sortByDate: SortByDate = (x) =>
  x.sort((a, b) => Number(a.date.toJSDate()) - Number(b.date.toJSDate()));
