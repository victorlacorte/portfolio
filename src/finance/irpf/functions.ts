import { div, mul } from 'src/utils/number';
import type { Position, PositionEntry } from 'src/types';

import { buyMessage, sellMessage } from './messages';

// sort descending to remove subsequent entries
// year and year - 1
export function relevantEntries(
  year: number,
  entries: PositionEntry[],
): PositionEntry[] {
  const filtered = entries
    .filter((e) => e.date.year <= year)
    .sort((a, b) => Number(b.date.toJSDate()) - Number(a.date.toJSDate()));

  const relevant: PositionEntry[] = [];

  // for (let i = 0; i < relevant.length; i++) {
  //   const curr = relevant[i];

  //   if (curr.quantity == 0 && curr.date.year <= year - 1) {
  //     return ret;
  //   }

  //   ret.push(curr);
  // }
  for (const e of filtered) {
    if (e.quantity == 0 && e.date.year <= year - 1) {
      return relevant.filter((e) => e.date.year == year);
    }

    relevant.push(e);
  }

  return relevant;
}

// const mergeEntries = (entries: PositionEntry[]) => {
//   const sorted = entries.sort(
//     (a, b) => Number(a.date.toJSDate) - Number(b.date.toJSDate()),
//   );
//   const ret = [];

//   for (let i = 0; i < sorted.length - 1; i++) {
//     if (
//       sorted[i].date.equals(sorted[i + 1].date) &&
//       ((sorted[i].buyQuantity && sorted[i + 1].buyQuantity) ||
//         (sorted[i].sellQuantity && sorted[i + 1].sellQuantity))
//     ) {
//       if (sorted[i].buyQuantity) {
//         // const [popped0] = sorted.splice(i, 1);
//         // const [popped1] = sorted.splice(i, 1); // splice mutates the array so i calc also changes

//       } else {
//       }
//       const popped = sorted;
//       i++;
//     }
//   }
// };

export function formatEntries(
  entries: PositionEntry[],
  separator = '\n',
): string {
  return entries
    .map((e) =>
      e.buyQuantity
        ? buyMessage({
            date: e.date,
            price: div(e.buyTotal, e.buyQuantity),
            quantity: e.buyQuantity,
          })
        : sellMessage({
            date: e.date,
            price: div(e.sellTotal, e.sellQuantity),
            profit: e.sellProfit,
            quantity: e.sellQuantity,
          }),
    )
    .join(separator);
}

/**
 * Sort entries descending and return the first one that satisfies
 * entry.date.year <= year or `undefined` if none are found
 *
 * @param year
 * @param entries
 */
export function lastEntryFromYear(
  year: number,
  entries: PositionEntry[],
): PositionEntry | undefined {
  const sorted = entries.sort(
    (a, b) => Number(b.date.toJSDate()) - Number(a.date.toJSDate()),
  );

  for (const e of sorted) {
    if (e.date.year <= year) {
      return e;
    }
  }
}
// Obtain:
// Value in Y - 1
// Value in Y
// Amount owned in Y
export function irpf(year: number, entries: PositionEntry[]) {
  const relevant = relevantEntries(year, entries);
  const lastFromYear = lastEntryFromYear(year, relevant);
  const lastFromPrevYear = lastEntryFromYear(year - 1, relevant);

  return {
    [year]: lastFromYear
      ? `${lastFromYear.quantity}, ${mul(
          lastFromYear.price,
          lastFromYear.quantity,
        )}`
      : 0,
    [year - 1]: lastFromPrevYear
      ? mul(lastFromPrevYear.price, lastFromPrevYear.quantity)
      : 0,
    log: formatEntries(
      relevant
        // .filter((e) => e.date.year == year)
        .sort((a, b) => Number(a.date.toJSDate()) - Number(b.date.toJSDate())),
    ),
  };
}
