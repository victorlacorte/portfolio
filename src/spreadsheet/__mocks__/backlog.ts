// Year, Month, Day, Ticker, Operation, Quantity, Price, Tax, IRRF
export const years = [[2020, 2020, 2020]];
export const months = [[1, 1, 1]];
export const days = [[1, 1, 1]];
export const tickers = [['foo1', 'foo1', 'foo1']];
export const operations = [['buy', 'sell', 'buy']];
export const quantities = [[200, 100, 100]];
export const prices = [[1, 2, 0.5]];
export const taxes = [[1, 1, 1]];
export const irrfs = [[0, 1.5, 0]];
