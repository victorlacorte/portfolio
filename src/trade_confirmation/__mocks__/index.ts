/**
 * 1 operation
 * 2 equal operations
 * 2 different operations
 */
