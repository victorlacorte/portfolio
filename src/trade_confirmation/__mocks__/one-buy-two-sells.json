{
  "Sheet1": [
    [
      "CLIENTE:",
      "123456789",
      "JOHN DOE",
      "DATA PREGÃO",
      "30062021",
      "NUM NOTA",
      "123456789",
      "",
      ""
    ],
    ["", "", "", "", "", "", "", "", ""],
    [
      "PRAÇA",
      "C/V",
      "TIPO DE MERCADO",
      "ESPECIFICAÇÃO DO TÍTULO",
      "OBS(*)",
      "QUANTIDADE",
      "PREÇO DE LIQUIDAÇÃO(R$)",
      "COMPRA/VENDA (R$)",
      "D/C"
    ],
    [
      "1-Bovespa",
      "C",
      "VIS",
      "AAAA3 ON      NM",
      "",
      "400",
      "5,75",
      "2.300,00",
      "D"
    ],
    ["", "", "", "SUBTOTAL:", "", "400", "5,7500", "2.300,00", ""],
    [
      "1-Bovespa",
      "V",
      "VIS",
      "BBBB4 PN      N1",
      "",
      "600",
      "11,18",
      "6.708,00",
      "C"
    ],
    [
      "1-Bovespa",
      "V",
      "VIS",
      "BBBB4 PN      N1",
      "",
      "600",
      "11,21",
      "6.726,00",
      "C"
    ],
    ["", "", "", "SUBTOTAL:", "", "1.200", "11,1950", "13.434,00", ""],
    [
      "1-Bovespa",
      "V",
      "VIS",
      "CCCC3 ON      N1",
      "",
      "300",
      "1,51",
      "453,00",
      "C"
    ],
    ["", "", "", "SUBTOTAL:", "", "300", "1,5100", "453,00", ""],
    ["", "", "", "", "", "", "", "", ""],
    ["RESUMO DOS NEGÓCIOS", "", "", "", "", "RESUMO FINANCEIRO", "", "", ""],
    [
      "DEBÊNTURES:",
      "0.00",
      "VENDAS À VISTA:",
      "13887.00",
      "",
      "VALOR LÍQ. DAS OPER.(1):",
      "11587.00 C",
      "TAXA DE LIQUIDAÇÃO(2):",
      "4.04 D"
    ],
    [
      "COMPRAS À VISTA:",
      "2300.00",
      "OPÇÕES - COMPRA:",
      "0.00",
      "",
      "TAXA REG.(3):",
      "0.00 D",
      "TOTAL(1+2+3)A:",
      "11582.96 C"
    ],
    [
      "OPÇÕES - VENDA:",
      "0.00",
      "OPER. À TERMO:",
      "0.00",
      "",
      "TX. TER/OPC/FUT:",
      "0.00 D",
      "TX. A.N.A.:",
      "0.00 D"
    ],
    [
      "OPER. FUT.:",
      "0.00",
      "VLR. OPER. C/ TIT. PUB.:",
      "0.00",
      "",
      "EMOLUMENTOS:",
      "0.80 D",
      "TOTAL BOLSA B:",
      "-0.80 D"
    ],
    [
      "VLR. OPER.:",
      "16187.00",
      "VLR. AJU. P/ FUT.:",
      "0.00",
      "",
      "CORRETAGEM:",
      "0.00 D",
      "ISS:",
      "0.00 D"
    ],
    [
      "IR S/ CORRETAGEM:",
      "0.00",
      "IRRF S/ DAY TRADE:",
      "0.00",
      "",
      "IRRF S/ OPER., BASE 13887.00:",
      "0.69 D",
      "OUTRAS:",
      "0,00 D"
    ],
    ["", "", "", "", "", "", "", "LIQ.(A+B) P/ 02/07/2021:", "11582.16 C"],
    []
  ]
}
