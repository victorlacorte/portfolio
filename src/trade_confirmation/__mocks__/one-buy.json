{
  "Sheet1": [
    [
      "CLIENTE:",
      "123456789",
      "JOHN DOE",
      "DATA PREGÃO",
      "05072021",
      "NUM NOTA",
      "123456789",
      "",
      ""
    ],
    ["", "", "", "", "", "", "", "", ""],
    [
      "PRAÇA",
      "C/V",
      "TIPO DE MERCADO",
      "ESPECIFICAÇÃO DO TÍTULO",
      "OBS(*)",
      "QUANTIDADE",
      "PREÇO DE LIQUIDAÇÃO(R$)",
      "COMPRA/VENDA (R$)",
      "D/C"
    ],
    [
      "1-Bovespa",
      "C",
      "VIS",
      "AAAA4 PN  EJ  N1",
      "",
      "200",
      "25,01",
      "5.002,00",
      "D"
    ],
    ["", "", "", "SUBTOTAL:", "", "200", "25,0100", "5.002,00", ""],
    ["", "", "", "", "", "", "", "", ""],
    ["RESUMO DOS NEGÓCIOS", "", "", "", "", "RESUMO FINANCEIRO", "", "", ""],
    [
      "DEBÊNTURES:",
      "0.00",
      "VENDAS À VISTA:",
      "0.00",
      "",
      "VALOR LÍQ. DAS OPER.(1):",
      "-5002.00 D",
      "TAXA DE LIQUIDAÇÃO(2):",
      "1.25 D"
    ],
    [
      "COMPRAS À VISTA:",
      "5002.00",
      "OPÇÕES - COMPRA:",
      "0.00",
      "",
      "TAXA REG.(3):",
      "0.00 D",
      "TOTAL(1+2+3)A:",
      "-5003.25 D"
    ],
    [
      "OPÇÕES - VENDA:",
      "0.00",
      "OPER. À TERMO:",
      "0.00",
      "",
      "TX. TER/OPC/FUT:",
      "0.00 D",
      "TX. A.N.A.:",
      "0.00 D"
    ],
    [
      "OPER. FUT.:",
      "0.00",
      "VLR. OPER. C/ TIT. PUB.:",
      "0.00",
      "",
      "EMOLUMENTOS:",
      "0.35 D",
      "TOTAL BOLSA B:",
      "-0.35 D"
    ],
    [
      "VLR. OPER.:",
      "5002.00",
      "VLR. AJU. P/ FUT.:",
      "0.00",
      "",
      "CORRETAGEM:",
      "0.00 D",
      "ISS:",
      "0.00 D"
    ],
    [
      "IR S/ CORRETAGEM:",
      "0.00",
      "IRRF S/ DAY TRADE:",
      "0.00",
      "",
      "IRRF S/ OPER., BASE 0.00:",
      "0.00 D",
      "OUTRAS:",
      "0,00 D"
    ],
    ["", "", "", "", "", "", "", "LIQ.(A+B) P/ 07/07/2021:", "-5003.60 D"],
    []
  ]
}
