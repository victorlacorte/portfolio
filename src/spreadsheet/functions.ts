import { statsFrom } from 'src/finance';
import type { BuySellEvent, Transaction } from 'src/types';

import Logger from '../utils/logging';
import { mul, sub } from '../utils/number';
import { expectedPositive } from '../utils/messages';
import { profitMessage, snapshotMessage } from './logging';

type ProfitReturn = {
  total: number;
  profit: number;
  taxDeduction: number;
  log: string;
};

// `total` and `profit` return properties should be changed
export function profit(
  year: number,
  month: number,
  transactions: Transaction[],
): ProfitReturn {
  let total = 0;
  let profit = 0;
  let taxDeduction = 0;
  const logger = new Logger();

  // `taxDeduction` is not part of the profit calculation
  const handleSell: BuySellEvent = (t, stats) => {
    if (t.taxDeduction < 0) {
      throw new Error(expectedPositive(t.taxDeduction));
    }

    if (t.date.year === year && t.date.month === month) {
      const purchasedValue = mul(stats[t.ticker].averagePrice, t.quantity);
      const currProfit = sub(t.total, purchasedValue);

      total += t.total;
      profit += currProfit;
      taxDeduction += t.taxDeduction;

      logger.add(
        profitMessage({
          ...t,
          purchasedValue,
          profit: currProfit,
        }),
      );
    }
  };

  statsFrom({
    transactions,
    onSell: handleSell,
  });

  return { total, profit, taxDeduction, log: logger.join() };
}

export function snapshot(transactions: Transaction[]) {
  const control: { [ticker: string]: Logger } = {};

  const handleOperation: BuySellEvent = (t, stats) => {
    if (!Object.prototype.hasOwnProperty.call(control, t.ticker)) {
      control[t.ticker] = new Logger();
    }

    control[t.ticker].add(snapshotMessage({ ...t }));
  };

  const stats = statsFrom({
    transactions,
    onBuy: handleOperation,
    onSell: handleOperation,
  });

  return Object.keys(control).map((ticker) => {
    if (Object.prototype.hasOwnProperty.call(stats, ticker)) {
      return {
        ticker,
        purchasedQuantity: stats[ticker].purchased.quantity,
        purchasedTotal: stats[ticker].purchased.total,
        soldQuantity: stats[ticker].sold.quantity,
        soldTotal: stats[ticker].sold.total,
        averagePrice: stats[ticker].averagePrice,
        log: control[ticker].join(),
      };
    }
  });
}
