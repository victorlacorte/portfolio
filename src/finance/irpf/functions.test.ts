import { makeBuyEntry, makeSellEntry, sortByDate, year } from './__mocks__';
import { formatEntries, lastEntryFromYear, relevantEntries } from './functions';

describe('finance/irpf/functions', () => {
  describe('relevantEntries', () => {
    test('happy flow: all entries are relevant', () => {
      expect(relevantEntries(year, [])).toEqual([]);

      const entries = Array.from({ length: 3 }, (_, index) =>
        makeBuyEntry({ day: index + 1 }),
      );

      expect(sortByDate(relevantEntries(year, entries))).toEqual(
        sortByDate(entries),
      );
    });

    test('some entries are excluded', () => {
      const entries = Array.from({ length: 3 }, (_, index) =>
        makeBuyEntry({ year: year + index }),
      );

      expect(relevantEntries(year, entries)).toEqual([entries[0]]);

      expect(sortByDate(relevantEntries(year + 1, entries))).toEqual(
        sortByDate(entries.slice(0, -1)),
      );
    });

    test('closed position entries prior to (and including) year - 1 are excluded', () => {
      const entries = [
        makeBuyEntry({ year: year - 2 }),
        makeBuyEntry({ year: year - 2 }),
        makeBuyEntry({ year: year - 2 }),
        makeSellEntry({ year: year - 1, quantity: 0 }),
        makeBuyEntry(),
      ];

      expect(relevantEntries(year, entries)).toEqual([
        entries[entries.length - 1],
      ]);

      expect(sortByDate(relevantEntries(year - 1, entries))).toEqual(
        sortByDate(entries.slice(0, -1)),
      );
    });
  });

  describe('lastEntryFromYear', () => {
    test('works as expected', () => {
      expect(lastEntryFromYear(year, [])).toBeUndefined();
      expect(lastEntryFromYear(year + 1, [])).toBeUndefined();

      const entries = [
        makeBuyEntry({ year: 2020 }),
        makeBuyEntry({ year: 2021 }),
        makeBuyEntry({ year: 2022 }),
      ];

      [2020, 2021, 2022].forEach((y) => {
        expect(lastEntryFromYear(y, entries)).toEqual(
          makeBuyEntry({ year: y }),
        );
      });
    });
  });

  describe('formatEntries', () => {
    test('works as expected', () => {
      const entries = [makeBuyEntry(), makeSellEntry()];
      expect(formatEntries(entries)).toMatchSnapshot();
    });
  });
});
