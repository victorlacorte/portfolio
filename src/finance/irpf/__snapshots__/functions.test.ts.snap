// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`finance/irpf/functions formatEntries works as expected 1`] = `
"01/01: 100 compradas, R$1.00/ação
01/01: 100 vendidas, R$1.00/ação, lucro de R$1.00"
`;
